R/changeLegendColour.R

```diff
@@ -20,69 +20,35 @@ changeLegendColour <- function(x,
 
   graph_grob <- ggplot2::ggplotGrob(x)
 
-  ##Build plot to extract deets from it
-  g <- ggplot2::ggplot_build(x)
-
-  if(is.null(g$plot$scales$scales[[1]]$palette.cache)) {
-    stop("You must choose options for the colour/fill in your ggplot2 object.")
-  }
-
-  if(is.null(base::names(g$plot$scales$scales[[1]]$palette.cache))) {
-    stop("You must provide names to the vector containing the colours you chose
-         which correspond to the group which is that colour:
-         scale_colour_manual(values = c('group 1' = 'colour 1', ...))")
-  }
-
-  if(is.null(base::names(g$plot$scales$scales[[1]]$palette.cache))) {
-    stop("You must provide names to the vector containing the colours you chose
-         which correspond to the group which is that colour:
-         scale_colour_manual(values = c('group 1' = 'colour 1', ...))")
-  }
-
-  if(base::length(g$plot$scales$scales[[1]]$labels) > 0 & is.null(base::names(g$plot$scales$scales[[1]]$labels))) {
-    stop("You must provide names to the vector containing the new labels you chose for your legend
-         which correspond to the group which has that new label:
-         scale_colour_manual(values = c('group 1' = 'label 1', ...))")
-  }
-  ##Extract names and colours automagically from plot if available from palette cache
-
-  leg_font_colour <- base::data.frame(
-    leg_text = base::names(g$plot$scales$scales[[1]]$palette.cache),
-    col = base::unname(g$plot$scales$scales[[1]]$palette.cache)
-  )
-
-  if(base::length(g$plot$scales$scales[[1]]$labels) > 0) {
-    new_leg_text <- base::data.frame(
-      leg_text = base::names(g$plot$scales$scales[[1]]$labels),
-      new_leg_text = base::unname(g$plot$scales$scales[[1]]$labels)
-    )
-
-    leg_font_colour <- base::merge(leg_font_colour, new_leg_text, by = "leg_text", all.x = TRUE)
-    leg_font_colour$new_leg_text[is.na(leg_font_colour$new_leg_text)] <- leg_font_colour$leg_text[is.na(leg_font_colour$new_leg_text)]
-    colnames(leg_font_colour)[colnames(leg_font_colour) == "leg_text"] <- "old_leg_text"
-    colnames(leg_font_colour)[colnames(leg_font_colour) == "new_leg_text"] <- "leg_text"
-  }
-
   #Find all the legend text
   gb <- base::grep("guide-box",
-                   graph_grob$layout$name)
+                   graph_grob$layout$name,
+                   fixed = TRUE)
+
   gb2 <- base::grep("guides",
-                    graph_grob$grobs[[gb]]$layout$name)
+                    graph_grob$grobs[[gb]]$layout$name,
+                    fixed = TRUE)
+
   label_text <- base::grep("label",
-                           graph_grob$grobs[[gb]]$grobs[[gb2]]$layout$name)
+                           graph_grob$grobs[[gb]]$grobs[[gb2]]$layout$name,
+                           fixed = TRUE)
+
+  ##Find colours associated with each legend object
+  ##Grobs of colour blobs
+  col_blobs <- grep("GRID.points",
```

Comment:
This could be fixed with a more generic grep? Not keen to go back to ggplot build as it fails when there's an unnamed selection of colours (e.g. in very simple charts).

Reply:
I will start by adding some unit testing for what already does work, as this is proving complicated

Reply:
That's a fair point. I'll take a look through the grob see what's around

Reply:
Replace looking for GRID.points with the below which is based on the names of the layout grobs. Should be more robust....hopefully

#Index of grobs with key in the name
keys <- grep("key", graph_grob$grobs[[gb]]$grobs[[gb2]]$layout$name, fixed = TRUE)

#Index of the above index without -bg
keys_no_bg <- grep("-bg", graph_grob$grobs[[gb]]$grobs[[gb2]]$layout$name[keys], fixed = TRUE, invert = TRUE)

#Index of grobs with key in the name but doesn't end -bg
col_blobs <- keys[keys_no_bg]